[delinearize] use SCEV exprs in getIndexExpressionsFromGEP

sebpop opened the pull request:
clean up interface of getIndexExpressionsFromGEP to get SCEV expressions instead of int for Sizes of the arrays.
This intends to simplify the code in #156342 by avoiding conversions from SCEV to int and back to SCEV.

sebpop committed:
[delinearize] use SCEV exprs in getIndexExpressionsFromGEP

kasuga-fj reviewed (commented):
Generally LGTM. Please apply clang-format.

sjoerdmeijer reviewed (approved):
With @kasuga-fj's minor comments addressed, LGTM too.

sebpop committed:
[delinearize] use DataLayout index type in getIndexExpressionsFromGEP

Use SE.getEffectiveSCEVType(GEP->getPointerOperandType()) to get the
correct pointer index sized integer type for the SCEV constant sizes,
instead of using the type of the expression. This ensures the sizes
have the correct type for the target's address space.
